## Chapter 5 – Exercise 1: Sử dụng ibm_db API trong Python để kết nối đến IBM DB2 warehouse

#### Học viên thực hiện các yêu cầu sau và đối chiếu với kết quả cho trước:

In [1]:
# {
#   "db": "BLUDB",
#   "dsn": "DATABASE=BLUDB;HOSTNAME=dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net;PORT=50000;PROTOCOL=TCPIP;UID=vhq84303;PWD=bq58n7p4sr4x@593;",
#   "host": "dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net",
#   "hostname": "dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net",
#   "https_url": "https://dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net",
#   "jdbcurl": "jdbc:db2://dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net:50000/BLUDB",
#   "parameters": {},
#   "password": "bq58n7p4sr4x@593",
#   "port": 50000,
#   "ssldsn": "DATABASE=BLUDB;HOSTNAME=dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net;PORT=50001;PROTOCOL=TCPIP;UID=vhq84303;PWD=bq58n7p4sr4x@593;Security=SSL;",
#   "ssljdbcurl": "jdbc:db2://dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net:50001/BLUDB:sslConnection=true;",
#   "uri": "db2://vhq84303:bq58n7p4sr4x%40593@dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net:50000/BLUDB",
#   "username": "vhq84303"
# }

In [2]:
import pandas as pd
import ibm_db_dbi

In [3]:
dsn_database = "BLUDB"            
dsn_hostname = "dashdb-txn-sbox-yp-dal09-12.services.dal.bluemix.net" 
dsn_port = "50000"                
dsn_protocol = "TCPIP"            
dsn_uid = "vhq84303"        
dsn_pwd = "bq58n7p4sr4x@593"

In [4]:
# Câu 1: Tạo kết nối đến Database QuanLySinhVien đã được tạo trong Chapter2_Exercise2
dsn = (
    "DATABASE={0};"
    "HOSTNAME={1};"
    "PORT={2};"
    "PROTOCOL={3};"
    "UID={4};"
    "PWD={5};").format(dsn_database, dsn_hostname, dsn_port, dsn_protocol, dsn_uid, dsn_pwd)
conn = ibm_db_dbi.connect(dsn, "", "")
print ("Connected!")

Connected!


In [5]:
stmt = 'select "Ma_sinh_vien", "Ho", "Ten", "Ngay_sinh", "Hoc_bong", "Ma_khoa" from sinh_vien'
df = pd.read_sql(stmt, conn)
conn.close()

True

In [6]:
df.head()

Ma_sinh_vien              Ho    Ten  Ngay_sinh  Hoc_bong Ma_khoa
0        C0004  Đinh Thị Thanh   Dung 1999-01-19   1102500      CN
1        C0007          Lê Thị    Mai 1999-09-13   2205000      CN
2        C0009  Thái Thị Thanh  Nhung 1999-07-18    500000      CN
3        C0011      Hoàng Thảo   Thúy 1999-12-07   1102500      CN
4        C0013     Nguyễn Minh    Thu 1999-07-18   1102500      CN

In [7]:
# Câu 2: Sử dụng ibm_db API, truy vấn dữ liệu trên 1 bảng

In [8]:
# Câu 2a: Liệt kê danh sách sinh viên gồm các cột Mã, Họ và Tên, Ngày sinh, Giới tính, Học bổng. Có sắp tăng theo cột học bổng.
def connect(dsn):
    conn = ibm_db_dbi.connect(dsn, "", "")
    return conn

In [9]:
conn = connect(dsn)
stmt = '''
SELECT "Ma_sinh_vien", "Ho"||' '||"Ten" AS "Ho_ten", 
CASE 
	WHEN "Gioi_tinh" THEN 'Nam' 
	ELSE 'Nữ'
END AS "Gioi_tinh", 
VARCHAR_FORMAT("Ngay_sinh", 'DD/MM/YYYY HH24:MI:SS') AS "Ngay_sinh",
VARCHAR_FORMAT("Hoc_bong",'9,999,999.99') AS "Hoc_bong_"
FROM "SINH_VIEN"
ORDER BY "Hoc_bong"
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_sinh_vien             Ho_ten Gioi_tinh            Ngay_sinh  \
0        T0507       Cao Minh Thu        Nữ  04/05/1999 00:00:00   
1        T0509       Văn Phú Minh       Nam  26/02/1999 00:00:00   
2        T0510   Khương Vĩnh Dũng       Nam  17/04/1999 00:00:00   
3        T0512       Đàm Vĩnh Huy       Nam  18/03/1999 00:00:00   
4        T0520  Bạch Văn Minh Lộc       Nam  19/09/1999 00:00:00   

       Hoc_bong_  
0            .00  
1            .00  
2            .00  
3            .00  
4            .00

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
   <pre> 5 dòng đầu của dữ liệu:</pre>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Gioi_tinh</th>
      <th>Ngay_sinh</th>
      <th>Hoc_bong_</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>T0507</td>
      <td>Cao Minh Thu</td>
      <td>Nữ</td>
      <td>04/05/1999 00:00:00</td>
      <td>.00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>T0509</td>
      <td>Văn Phú Minh</td>
      <td>Nam</td>
      <td>26/02/1999 00:00:00</td>
      <td>.00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>T0510</td>
      <td>Khương Vĩnh Dũng</td>
      <td>Nam</td>
      <td>17/04/1999 00:00:00</td>
      <td>.00</td>
    </tr>
    <tr>
      <th>3</th>
      <td>T0512</td>
      <td>Đàm Vĩnh Huy</td>
      <td>Nam</td>
      <td>18/03/1999 00:00:00</td>
      <td>.00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>T0520</td>
      <td>Bạch Văn Minh Lộc</td>
      <td>Nam</td>
      <td>19/09/1999 00:00:00</td>
      <td>.00</td>
    </tr>
  </tbody>
</table>
</div>

<pre> 5 dòng cuối của dữ liệu:</pre>
<div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Gioi_tinh</th>
      <th>Ngay_sinh</th>
      <th>Hoc_bong_</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>730</th>
      <td>C0461</td>
      <td>Hoàng Thảo Hạnh</td>
      <td>Nữ</td>
      <td>21/02/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
    <tr>
      <th>731</th>
      <td>C0466</td>
      <td>Lâm Thảo Hoa</td>
      <td>Nữ</td>
      <td>18/04/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
    <tr>
      <th>732</th>
      <td>C0476</td>
      <td>Cao Thị Thảo Trang</td>
      <td>Nữ</td>
      <td>04/01/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
    <tr>
      <th>733</th>
      <td>C0481</td>
      <td>Bạch Minh Thu</td>
      <td>Nữ</td>
      <td>24/02/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
    <tr>
      <th>734</th>
      <td>C0493</td>
      <td>Đinh Phú Mỹ</td>
      <td>Nữ</td>
      <td>06/02/1999 00:00:00</td>
      <td>3,307,500.00</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [10]:
df.tail()

Ma_sinh_vien              Ho_ten Gioi_tinh            Ngay_sinh  \
730        C0461     Hoàng Thảo Hạnh        Nữ  21/02/1999 00:00:00   
731        C0466        Lâm Thảo Hoa        Nữ  18/04/1999 00:00:00   
732        C0476  Cao Thị Thảo Trang        Nữ  04/01/1999 00:00:00   
733        C0481       Bạch Minh Thu        Nữ  24/02/1999 00:00:00   
734        C0493         Đinh Phú Mỹ        Nữ  06/02/1999 00:00:00   

         Hoc_bong_  
730   3,307,500.00  
731   3,307,500.00  
732   3,307,500.00  
733   3,307,500.00  
734   3,307,500.00

In [11]:
# Câu 2b: Hiển thị danh sách 10 sinh viên đầu tiên có học bổng từ cao đến thấp 
# Hướng dẫn: dùng FETCH FIRST 10 ROWS ONLY và ORDER BY … DESC
conn = connect(dsn)
stmt = '''
SELECT *
  FROM "SINH_VIEN"
  WHERE "Hoc_bong">0
  ORDER BY "Hoc_bong" DESC
  FETCH FIRST 10 ROWS ONLY;
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_sinh_vien           Ho     Ten Gioi_tinh  Ngay_sinh                Mail  \
0        C0025  Phạm Phương  Phương     False 1999-06-15  ppphuong@gmail.com   
1        C0080   Bạch Thanh   Thanh     False 1999-01-04   btthanh@gmail.com   
2        C0075      Lê Thảo     Hoa     False 1999-05-15     lthoa@gmail.com   
3        C0074    Thái Thảo     Nga     False 1999-02-23     ttnga@gmail.com   
4        C0065     Lâm Thảo    Trúc     False 1999-09-19    lttruc@gmail.com   

      Di_dong          CMND  Hoc_bong Ma_khoa  
0  0168374535  774375787574   3307500      CN  
1  0913843497  169614465950   3307500      CN  
2  0913243932  662273790256   3307500      CN  
3  0168122579  443293205719   3307500      CN  
4  0987113162  613542186162   3307500      CN

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho</th>
      <th>Ten</th>
      <th>Gioi_tinh</th>
      <th>Ngay_sinh</th>
      <th>Mail</th>
      <th>Di_dong</th>
      <th>CMND</th>
      <th>Hoc_bong</th>
      <th>Ma_khoa</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0025</td>
      <td>Phạm Phương</td>
      <td>Phương</td>
      <td>False</td>
      <td>1999-06-15</td>
      <td>ppphuong@gmail.com</td>
      <td>0168374535</td>
      <td>774375787574</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0080</td>
      <td>Bạch Thanh</td>
      <td>Thanh</td>
      <td>False</td>
      <td>1999-01-04</td>
      <td>btthanh@gmail.com</td>
      <td>0913843497</td>
      <td>169614465950</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0075</td>
      <td>Lê Thảo</td>
      <td>Hoa</td>
      <td>False</td>
      <td>1999-05-15</td>
      <td>lthoa@gmail.com</td>
      <td>0913243932</td>
      <td>662273790256</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0074</td>
      <td>Thái Thảo</td>
      <td>Nga</td>
      <td>False</td>
      <td>1999-02-23</td>
      <td>ttnga@gmail.com</td>
      <td>0168122579</td>
      <td>443293205719</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0065</td>
      <td>Lâm Thảo</td>
      <td>Trúc</td>
      <td>False</td>
      <td>1999-09-19</td>
      <td>lttruc@gmail.com</td>
      <td>0987113162</td>
      <td>613542186162</td>
      <td>3307500</td>
      <td>CN</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [12]:
# Câu 3: Sử dụng ibm_db API, truy vấn trên nhiều bảng

In [13]:
# Câu 3a: Liệt kê danh sách sinh viên gồm các cột mã, họ, tên, ngày sinh, giới tính, mã khoa, tên khoa
conn = connect(dsn)
stmt = '''
SELECT s."Ma_sinh_vien", s."Ho", s."Ten", s."Ngay_sinh", s."Gioi_tinh", s."Ma_khoa", k."Ten" AS "Ten_khoa"
  FROM "SINH_VIEN" s INNER JOIN "KHOA" k
  ON s."Ma_khoa"=k."Ma_khoa"
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_sinh_vien              Ho    Ten  Ngay_sinh Gioi_tinh Ma_khoa  \
0        T0504  Đặng Thị Thanh    Lan 1999-02-22     False      TO   
1        C0004  Đinh Thị Thanh   Dung 1999-01-19     False      CN   
2        C0007          Lê Thị    Mai 1999-09-13     False      CN   
3        C0009  Thái Thị Thanh  Nhung 1999-07-18     False      CN   
4        C0011      Hoàng Thảo   Thúy 1999-12-07     False      CN   

                   Ten_khoa  
0                 Khoa Toán  
1  Khoa Công nghệ Thông tin  
2  Khoa Công nghệ Thông tin  
3  Khoa Công nghệ Thông tin  
4  Khoa Công nghệ Thông tin

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho</th>
      <th>Ten</th>
      <th>Ngay_sinh</th>
      <th>Gioi_tinh</th>
      <th>Ma_khoa</th>
      <th>Ten_khoa</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>T0504</td>
      <td>Đặng Thị Thanh</td>
      <td>Lan</td>
      <td>1999-02-22</td>
      <td>False</td>
      <td>TO</td>
      <td>Khoa Toán</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh</td>
      <td>Dung</td>
      <td>1999-01-19</td>
      <td>False</td>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0007</td>
      <td>Lê Thị</td>
      <td>Mai</td>
      <td>1999-09-13</td>
      <td>False</td>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0009</td>
      <td>Thái Thị Thanh</td>
      <td>Nhung</td>
      <td>1999-07-18</td>
      <td>False</td>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0011</td>
      <td>Hoàng Thảo</td>
      <td>Thúy</td>
      <td>1999-12-07</td>
      <td>False</td>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [14]:
# Câu 3b: Liệt kê kết quả học tập gồm các cột mã sinh viên, họ và tên, mã môn học, điểm
conn = connect(dsn)
stmt = '''
SELECT s."Ma_sinh_vien", s."Ho"||' '||s."Ten" AS "Ho_ten", k."Ma_mon", k."Diem"
  FROM "SINH_VIEN" s INNER JOIN "KET_QUA" k ON s."Ma_sinh_vien"=k."Ma_sinh_vien";
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_sinh_vien               Ho_ten Ma_mon  Diem
0        C0092      Trần Thảo Nhung   LQL2   5.0
1        C0004  Đinh Thị Thanh Dung   CSD1   5.5
2        C0004  Đinh Thị Thanh Dung   CSD2   9.5
3        C0004  Đinh Thị Thanh Dung   CTDL   6.5
4        C0004  Đinh Thị Thanh Dung   KTLT   6.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ma_mon</th>
      <th>Diem</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0092</td>
      <td>Trần Thảo Nhung</td>
      <td>LQL2</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>CSD1</td>
      <td>5.5</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>CSD2</td>
      <td>9.5</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>CTDL</td>
      <td>6.5</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>KTLT</td>
      <td>6.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [15]:
# Câu 3c: Liệt kê danh sách các sinh viên (gồm các cột mã, họ tên, ngày sinh) có học môn ‘Cơ sở dữ liệu’
conn = connect(dsn)
stmt = '''
SELECT "Ma_sinh_vien", "Ho"||' '||"Ten" AS "Ho_ten", "Ngay_sinh"
FROM "SINH_VIEN"
WHERE "Ma_sinh_vien" IN (
SELECT k."Ma_sinh_vien" FROM "KET_QUA" k JOIN "MON_HOC" m ON k."Ma_mon"=m."Ma_mon"
WHERE m."Ten" LIKE 'Cơ sở dữ liệu%');
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_sinh_vien                Ho_ten  Ngay_sinh
0        C0445          Đàm Ngọc Thu 1999-09-18
1        C0004   Đinh Thị Thanh Dung 1999-01-19
2        C0007            Lê Thị Mai 1999-09-13
3        C0009  Thái Thị Thanh Nhung 1999-07-18
4        C0011       Hoàng Thảo Thúy 1999-12-07

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ngay_sinh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0445</td>
      <td>Đàm Ngọc Thu</td>
      <td>1999-09-18</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>1999-01-19</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>1999-09-13</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>1999-07-18</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>1999-12-07</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [16]:
# Câu 3d: Liệt kê danh sách các sinh viên (gồm các cột mã, họ tên, ngày sinh) đạt điểm 10 trong bài thi
conn = connect(dsn)
stmt = '''
SELECT "Ma_sinh_vien", "Ho"||' '||"Ten" AS "Ho_ten", "Ngay_sinh"
FROM "SINH_VIEN"
WHERE "Ma_sinh_vien" IN (SELECT "Ma_sinh_vien" FROM "KET_QUA" WHERE "Diem"=10);
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_sinh_vien                Ho_ten  Ngay_sinh
0        V0620           Lê Mạnh Bảo 1999-02-24
1        C0007            Lê Thị Mai 1999-09-13
2        C0009  Thái Thị Thanh Nhung 1999-07-18
3        C0011       Hoàng Thảo Thúy 1999-12-07
4        C0017       Lê Thị Thảo Đào 1999-12-27

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ngay_sinh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>V0620</td>
      <td>Lê Mạnh Bảo</td>
      <td>1999-02-24</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>1999-09-13</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>1999-07-18</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>1999-12-07</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0017</td>
      <td>Lê Thị Thảo Đào</td>
      <td>1999-12-27</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [17]:
# Câu 3e: Liệt kê danh sách các sinh viên (gồm các cột mã, họ tên, ngày sinh) có điểm thi môn ‘Toán cao cấp’ nhỏ hơn 5 
conn = connect(dsn)
stmt = '''
SELECT "Ma_sinh_vien", "Ho"||' '||"Ten" AS "Ho_ten", "Ngay_sinh"
FROM "SINH_VIEN"
WHERE "Ma_sinh_vien" IN (
SELECT k."Ma_sinh_vien" FROM "KET_QUA" k JOIN "MON_HOC" m ON k."Ma_mon"=m."Ma_mon"
WHERE m."Ten" LIKE 'Toán cao cấp%' AND k."Diem">5);
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_sinh_vien                Ho_ten  Ngay_sinh
0        C0442     Văn Thị Thanh Nga 1999-08-09
1        C0004   Đinh Thị Thanh Dung 1999-01-19
2        C0007            Lê Thị Mai 1999-09-13
3        C0009  Thái Thị Thanh Nhung 1999-07-18
4        C0011       Hoàng Thảo Thúy 1999-12-07

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ngay_sinh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0442</td>
      <td>Văn Thị Thanh Nga</td>
      <td>1999-08-09</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>1999-01-19</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>1999-09-13</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>1999-07-18</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>1999-12-07</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [18]:
# Câu 3f: Cho biết danh sách các môn học chưa có sinh viên đăng ký 
conn = connect(dsn)
stmt = '''
SELECT *
FROM "MON_HOC"
WHERE "Ma_mon" NOT IN (SELECT "Ma_mon" FROM "KET_QUA");
'''
df = pd.read_sql(stmt, conn)
df.head()

Empty DataFrame
Columns: [Ma_mon, Ten]
Index: []

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
    </tr>
  </thead>
  <tbody>
  </tbody>
</table>
</div>
  
</details>

In [19]:
# Câu 3g: Cho biết danh sách các sinh viên (gồm các cột mã, họ tên, ngày sinh) chưa đăng ký học môn học nào
conn = connect(dsn)
stmt = '''
SELECT "Ma_sinh_vien", "Ho"||' '||"Ten" AS "Ho_ten", "Ngay_sinh"
FROM "SINH_VIEN"
WHERE "Ma_sinh_vien" NOT IN (SELECT "Ma_sinh_vien" FROM "KET_QUA");
'''
df = pd.read_sql(stmt, conn)
df.head()

Empty DataFrame
Columns: [Ma_sinh_vien, Ho_ten, Ngay_sinh]
Index: []

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Ngay_sinh</th>
    </tr>
  </thead>
  <tbody>
  </tbody>
</table>
</div>
  
</details>

In [20]:
# Câu 4: Sử dụng ibm_db API, truy vấn có nhóm

In [21]:
# Câu 4a: Thống kê tổng số sinh viên theo khoa. Có sắp tăng theo tổng số sinh viên
conn = connect(dsn)
stmt = '''
SELECT "Ma_khoa", COUNT(*) AS "Tong_so_sinh_vien"
FROM "SINH_VIEN"
GROUP BY "Ma_khoa"
ORDER BY COUNT(*);
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_khoa  Tong_so_sinh_vien
0      TO              100.0
1      CN              235.0
2      VL              400.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_khoa</th>
      <th>Tong_so_sinh_vien</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TO</td>
      <td>100.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CN</td>
      <td>235.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VL</td>
      <td>400.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [22]:
# Câu 4b: Thống kê tổng học bổng theo khoa. Chỉ hiển thị những khoa có tổng học bổng >=3,000,000
conn = connect(dsn)
stmt = '''
SELECT s."Ma_khoa", k."Ten", SUM(s."Hoc_bong") AS "Tong_hoc_bong"
FROM "SINH_VIEN" s JOIN "KHOA" k ON s."Ma_khoa"=k."Ma_khoa"
GROUP BY s."Ma_khoa", k."Ten"
HAVING SUM(s."Hoc_bong")>=3000000;
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_khoa                       Ten  Tong_hoc_bong
0      CN  Khoa Công nghệ Thông tin      397325000
1      TO                 Khoa Toán      140000000
2      VL               Khoa Vật lý      616000000

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_khoa</th>
      <th>Ten</th>
      <th>Tong_hoc_bong</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
      <td>397325000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TO</td>
      <td>Khoa Toán</td>
      <td>140000000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VL</td>
      <td>Khoa Vật lý</td>
      <td>616000000</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [23]:
# Câu 4c: Hiển thị danh sách sinh viên gồm mã, họ tên và điểm trung bình các môn học mà sinh viên có đăng ký học
conn = connect(dsn)
stmt = '''
SELECT s."Ma_sinh_vien", s."Ho"||' '||s."Ten" AS "Ho_ten", AVG(k."Diem")
  FROM "SINH_VIEN" s INNER JOIN "KET_QUA" k ON s."Ma_sinh_vien"=k."Ma_sinh_vien"
GROUP BY s."Ma_sinh_vien", s."Ho"||' '||s."Ten";
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_sinh_vien                Ho_ten         3
0        C0004   Đinh Thị Thanh Dung  6.950000
1        C0007            Lê Thị Mai  7.700000
2        C0009  Thái Thị Thanh Nhung  7.400000
3        C0011       Hoàng Thảo Thúy  7.944444
4        C0013       Nguyễn Minh Thu  8.000000

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>6.950000</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>7.700000</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>7.400000</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>7.944444</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0013</td>
      <td>Nguyễn Minh Thu</td>
      <td>8.000000</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [24]:
# Câu 4d: Hiển thị danh sách khoa gồm mã khoa, tên khoa và điểm trung bình của tất cả các sinh viên trong khoa
conn = connect(dsn)
stmt = '''
SELECT s."Ma_khoa", k."Ten", AVG(kq."Diem") AS "Diem_trung_binh"
FROM "SINH_VIEN" s JOIN "KHOA" k ON s."Ma_khoa"=k."Ma_khoa"
	INNER JOIN "KET_QUA" kq ON s."Ma_sinh_vien"=kq."Ma_sinh_vien"
GROUP BY s."Ma_khoa", k."Ten";
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_khoa                       Ten  Diem_trung_binh
0      CN  Khoa Công nghệ Thông tin         7.544465
1      TO                 Khoa Toán         7.551948
2      VL               Khoa Vật lý         7.474493

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_khoa</th>
      <th>Ten</th>
      <th>Diem_trung_binh</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CN</td>
      <td>Khoa Công nghệ Thông tin</td>
      <td>7.544465</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TO</td>
      <td>Khoa Toán</td>
      <td>7.551948</td>
    </tr>
    <tr>
      <th>2</th>
      <td>VL</td>
      <td>Khoa Vật lý</td>
      <td>7.474493</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [25]:
# Câu 4e: Hiển thị danh sách sinh viên gồm mã, họ tên, điểm trung bình và kết quả xếp loại học tập của sinh viên, 
# trong đó kết quả xếp loại học tập được xét dựa trên điểm trung bình theo quy tắc sau: 
#                                Điểm trung bình < 5 :  Yếu
#                         5 <= Điểm trung bình < 6.5 :  Trung bình
#                         6.5 <= Điểm trung bình < 8 :  Khá
#                               Điểm trung bình >= 8 :  Giỏi
conn = connect(dsn)
stmt = '''
SELECT s."Ma_sinh_vien", s."Ho"||' '||s."Ten" AS "Ho_ten", AVG(k."Diem") AS "Diem_trung_binh",
  CASE 
	WHEN AVG(k."Diem")<5 THEN 'Yếu'
	WHEN AVG(k."Diem")<6.5 THEN 'Trung bình'
	WHEN AVG(k."Diem")<8 THEN 'Khá'
	ELSE 'Giỏi'
  END AS "Xep_loai"
  FROM "SINH_VIEN" s INNER JOIN "KET_QUA" k ON s."Ma_sinh_vien"=k."Ma_sinh_vien"
GROUP BY s."Ma_sinh_vien", s."Ho"||' '||s."Ten";
'''
df = pd.read_sql(stmt, conn)
df.head()


Ma_sinh_vien                Ho_ten  Diem_trung_binh Xep_loai
0        C0004   Đinh Thị Thanh Dung         6.950000      Khá
1        C0007            Lê Thị Mai         7.700000      Khá
2        C0009  Thái Thị Thanh Nhung         7.400000      Khá
3        C0011       Hoàng Thảo Thúy         7.944444      Khá
4        C0013       Nguyễn Minh Thu         8.000000     Giỏi

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_sinh_vien</th>
      <th>Ho_ten</th>
      <th>Diem_trung_binh</th>
      <th>Xep_loai</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>C0004</td>
      <td>Đinh Thị Thanh Dung</td>
      <td>6.950000</td>
      <td>Khá</td>
    </tr>
    <tr>
      <th>1</th>
      <td>C0007</td>
      <td>Lê Thị Mai</td>
      <td>7.700000</td>
      <td>Khá</td>
    </tr>
    <tr>
      <th>2</th>
      <td>C0009</td>
      <td>Thái Thị Thanh Nhung</td>
      <td>7.400000</td>
      <td>Khá</td>
    </tr>
    <tr>
      <th>3</th>
      <td>C0011</td>
      <td>Hoàng Thảo Thúy</td>
      <td>7.944444</td>
      <td>Khá</td>
    </tr>
    <tr>
      <th>4</th>
      <td>C0013</td>
      <td>Nguyễn Minh Thu</td>
      <td>8.000000</td>
      <td>Giỏi</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [26]:
# Câu 4f: Hiển thị danh sách các môn học cùng với điểm thi cao nhất mà sinh viên đã đạt được trong môn học đó, 
# kết quả trả về gồm mã môn, tên môn, điểm thi cao nhất của sinh viên 
conn = connect(dsn)
stmt = '''
SELECT m."Ma_mon", m."Ten", MAX(k."Diem") AS "Diem_thi_cao_nhat"
  FROM "MON_HOC" m INNER JOIN "KET_QUA" k ON m."Ma_mon"=k."Ma_mon"
GROUP BY m."Ma_mon", m."Ten";
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_mon                  Ten  Diem_thi_cao_nhat
0   CSD1      Cơ sở dữ liệu 1               10.0
1   CSD2      Cơ sở dữ liệu 2               10.0
2   CTDL     Cấu trúc dữ liệu               10.0
3   KTLT   Kỹ thuật Lập trình               10.0
4   LQL1  Lập trình Quản lý 1               10.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
      <th>Diem_thi_cao_nhat</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CSD1</td>
      <td>Cơ sở dữ liệu 1</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CSD2</td>
      <td>Cơ sở dữ liệu 2</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CTDL</td>
      <td>Cấu trúc dữ liệu</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KTLT</td>
      <td>Kỹ thuật Lập trình</td>
      <td>10.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>LQL1</td>
      <td>Lập trình Quản lý 1</td>
      <td>10.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [27]:
# Câu 4g: Hiển thị danh sách các môn học cùng với số lượng sinh viên đã đăng ký học các môn học đó, 
# kết quả trả về gồm mã môn, tên môn, số lượng sinh viên đã đăng ký 
conn = connect(dsn)
stmt = '''
SELECT m."Ma_mon", m."Ten", COUNT(k."Ma_sinh_vien") AS "So_luong_sinh_vien"
  FROM "MON_HOC" m INNER JOIN "KET_QUA" k ON m."Ma_mon"=k."Ma_mon"
GROUP BY m."Ma_mon", m."Ten";
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_mon                  Ten  So_luong_sinh_vien
0   CSD1      Cơ sở dữ liệu 1               930.0
1   CSD2      Cơ sở dữ liệu 2               921.0
2   CTDL     Cấu trúc dữ liệu               907.0
3   KTLT   Kỹ thuật Lập trình               919.0
4   LQL1  Lập trình Quản lý 1               928.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
      <th>So_luong_sinh_vien</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CSD1</td>
      <td>Cơ sở dữ liệu 1</td>
      <td>930.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CSD2</td>
      <td>Cơ sở dữ liệu 2</td>
      <td>921.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CTDL</td>
      <td>Cấu trúc dữ liệu</td>
      <td>907.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KTLT</td>
      <td>Kỹ thuật Lập trình</td>
      <td>919.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>LQL1</td>
      <td>Lập trình Quản lý 1</td>
      <td>928.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [28]:
# Câu 4h: Hiển thị danh sách các môn học có số lượng sinh viên đăng ký học đông nhất, 
# kết quả trả về gồm mã môn, tên môn và số lượng sinh viên đăng ký 
conn = connect(dsn)
stmt = '''
SELECT m."Ma_mon", m."Ten", COUNT(k."Ma_sinh_vien") AS "So_luong_sinh_vien"
  FROM "MON_HOC" m INNER JOIN "KET_QUA" k ON m."Ma_mon"=k."Ma_mon"
GROUP BY m."Ma_mon", m."Ten"
ORDER BY COUNT(k."Ma_sinh_vien") DESC
FETCH FIRST 1 ROWS ONLY;
'''
df = pd.read_sql(stmt, conn)
df.head()

Ma_mon              Ten  So_luong_sinh_vien
0   CSD1  Cơ sở dữ liệu 1               930.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
      <th>So_luong_sinh_vien</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CSD1</td>
      <td>Cơ sở dữ liệu 1</td>
      <td>930.0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [29]:
# Câu 4i: Thống kê số lượng sinh viên đậu và rớt của từng môn, biết rằng sinh viên rớt khi điểm thi nhỏ hơn 5, 
# kết quả trả về gồm mã môn, tên môn, số sinh viên đậu, số sinh viên rớt 
conn = connect(dsn)
stmt = '''
SELECT m."Ma_mon", m."Ten",
	SUM(CASE WHEN k."Diem">=5 THEN 1 ELSE 0 END) AS "Pass",
	SUM(CASE WHEN k."Diem"<5 THEN 1 ELSE 0 END) AS "Fail"
  FROM "MON_HOC" m INNER JOIN "KET_QUA" k ON m."Ma_mon"=k."Ma_mon"
GROUP BY m."Ma_mon", m."Ten";
'''
df = pd.read_sql(stmt, conn)
df

Ma_mon                  Ten  Pass  Fail
0   CSD1      Cơ sở dữ liệu 1   930     0
1   CSD2      Cơ sở dữ liệu 2   921     0
2   CTDL     Cấu trúc dữ liệu   907     0
3   KTLT   Kỹ thuật Lập trình   919     0
4   LQL1  Lập trình Quản lý 1   928     0
5   LQL2  Lập trình Quản lý 2   920     0
6   NMLT   Nhập môn Lập trình   912     0
7   TOA1      Toán cao cấp A1   911     0
8   TOA2      Toán cao cấp A2   911     0
9   TTNT     Trí tuệ nhân tạo   914     0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Ma_mon</th>
      <th>Ten</th>
      <th>Pass</th>
      <th>Fail</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CSD1</td>
      <td>Cơ sở dữ liệu 1</td>
      <td>930</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CSD2</td>
      <td>Cơ sở dữ liệu 2</td>
      <td>921</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CTDL</td>
      <td>Cấu trúc dữ liệu</td>
      <td>907</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KTLT</td>
      <td>Kỹ thuật Lập trình</td>
      <td>919</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>LQL1</td>
      <td>Lập trình Quản lý 1</td>
      <td>928</td>
      <td>0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>LQL2</td>
      <td>Lập trình Quản lý 2</td>
      <td>920</td>
      <td>0</td>
    </tr>
    <tr>
      <th>6</th>
      <td>NMLT</td>
      <td>Nhập môn Lập trình</td>
      <td>912</td>
      <td>0</td>
    </tr>
    <tr>
      <th>7</th>
      <td>TOA1</td>
      <td>Toán cao cấp A1</td>
      <td>911</td>
      <td>0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>TOA2</td>
      <td>Toán cao cấp A2</td>
      <td>911</td>
      <td>0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>TTNT</td>
      <td>Trí tuệ nhân tạo</td>
      <td>914</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [30]:
conn.close()

True